### PR TITLE
[Feature] Thrift support for realtime and batch ingestion

### PR DESCRIPTION
**Pull request by du00cs**
In our company thrift is used heavily, thrift serialized objects are stored to HDFS as SequenceFile and bytes to Kafka. I try to make it easier for thrift data to be ingested into druid without ETL in simple case.
     Thanks to the flexible JSONSpec, the Implementation is rather simple, deserialize thrift bytes to thrift object and then serialize it to json. And then parse it with JSONSpec.

Is anybody using thrift too? Hope this useful for somebody like me.

### CODE REVIEW

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  public ThriftInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("thriftJar") String jarPath,
+      @JsonProperty("thriftClass") String thriftClassName,
+      @JsonProperty("protocol") String protocol
+  )
+  {
+    this.protocol = protocol;
+
+    if (protocol == null || protocol.equals("compact")) {
```
Comment (fjy) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: can we use Jackson to deserialize the protocol into a TCompactProtocol or TBinaryProtocol? this iwll avoid the if/else check

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    this.jarPath = jarPath;
+    this.thriftClassName = thriftClassName;
+    Preconditions.checkNotNull(thriftClassName, "thrift class name");
+
+    this.parseSpec = parseSpec;
+    this.parser = parseSpec.makeParser();
+  }
+
+  protected Class<TBase> getThriftClass()
```
Comment (fjy) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: does this need to be protected?

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      throws IOException, ClassNotFoundException, IllegalAccessException, InstantiationException
+  {
+    final Class<TBase> thrift;
+    if(jarPath != null){
```
Comment (fjy) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: small formatting problem here, missing whitespace

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        bytes = ((BytesWritable) input).getBytes();
+      } else if (input instanceof ByteBuffer) {
+        bytes = ((ByteBuffer) input).array();
+      } else {
+        throw new IAE("unsupport input class of [%s]", input.getClass());
+      }
+
+      TBase o = thriftClass.newInstance();
+      deserializer.deserialize(o, bytes);
+      String json = serializer.toString(o);
```
Comment (fjy) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: if you are just converting thrift to Json, you dont need to pass in a parser, just use the Jsonparser?

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java (hunk 1 above)
Comment (du00cs) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: Any example? I am not familiar with Jackson and don't know how to make it after some search.

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java (hunk 4 above)
Comment (du00cs) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: I also want to support nested structure with JsonPath in JsonParseSpec(druid), Jsonparser(metamx java-util) is not enough.

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java (hunk 3 above)
Comment (du00cs) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: I will fix it.

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java (hunk 1 above)
Comment (du00cs) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftInputRowParser.java]: @fjy Sorry, but how to do that? Any tips?

**Comment by fjy** [comment]
@du00cs there's some overlap between this and https://github.com/druid-io/druid/pull/3633/files

Can we come to an agreement around which version to use?

**Comment by du00cs** [comment]
@fjy Protocol aware is awesome and I missed thread safe problem of deserializer. Something should be checked more carefully like how and when a thrift class is added to classpath. Implement a TBase to Map is too complicated and error prone. I should argue that using JSONPath will make it much simpler and more expressive and be consistent with what JsonSpec can do.

**Comment by justsmoke** [comment]
@du00cs I just had a look at your code, specify thrift jar in runtime is really good, and serialize to simple json can solve the problem of nested field, so I think your solution is better. Also, I noticed that your batch ingestion is based on SequenceFileInputFormat, can you help add support for LzoThriftBlockInputFormat (this is implemented in project elephant-bird which is also quite popular)

**Comment by du00cs** [comment]
@justsmoke I must admit that your protocol aware and taking thread safe into account is much better. We can do it together. Take multiple format into account should be supported.

**Comment by du00cs** [comment]
@justsmoke I have merged your work. Can you help me to test if LzoBlockThriftInputFormat is supported? I tried to set up the environment but failed.

**Comment by fjy** [comment]
@justsmoke @du00cs thanks, we can focus review on this PR

**Comment by du00cs** [comment]
@justsmoke

**Comment by du00cs** [comment]
@justsmoke update with a change to avoid code style check failure. Thrift generator is platform dependent, I use scrooge-generator as a substitution to generate thrift files. You may need to change the thrift and scrooge version to match your case.

**Comment by justsmoke** [comment]
@du00cs ok, I will test the LzoBlockThriftInputFormat and give you feedback in two days

**Comment by justsmoke** [comment]
@du00cs When using LzoBlockThriftInputFormat, the thrift jar had to be included in classpath. This is different from other cases, and you may need to add some description in documentation. After the thrift jar is added to classpath, the thrift object is successfully parsed.

**Comment by du00cs** [comment]
@justsmoke if the thrift jar is specified by `tmpjars` like 
```json
{
"jobProperties" : {
        "tmpjars": "/user/xxx/druid/test/book.jar"
      }
}
```
the job will succeed?

**Comment by justsmoke** [comment]
@du00cs No, using config: "elephantbird.class.for.MultiInputFormat" : "${YOUR_THRIFT_CLASS_NAME}" and include jar in classpath is the only way to get it work. (because thrift deserialize is done within the elephant-bird lib)

**Comment by du00cs** [comment]
@justsmoke A little confusing
```json
{
"jobProperties" : {
        "tmpjars": "/user/xxx/druid/test/book.jar",
        "elephantbird.class.for.MultiInputFormat" : "your.class.name"
      }
}
```
fails? I can not understand why.

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftDeserialization.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2016 XiaoMi Inc. All Rights Reserved.
```
Comment (gzm55) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftDeserialization.java]: this 'Copyright' is not appropriate here.

**Comment by justsmoke** [comment]
@du00cs Because deserialization is done by elephant-bird (after calling get on thriftWritable object, we get a TBase directly), so the parameter tmpjars is actually not used.

File: extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftDeserialization.java (hunk 5 above)
Comment (du00cs) [inline extensions-contrib/thrift-extensions/src/main/java/io/druid/data/input/thrift/ThriftDeserialization.java]: Sorry, it's just a mistake.

**Comment by fjy** [comment]
@du00cs @justsmoke there's a few merge conflicts. Do you guys feel this is ready?

**Comment by du00cs** [comment]
@justsmoke Can you send me a sample file of LzoBlockThriftInputFormat (and also the thrfit file) ? I want to look more into it.

**Comment by fjy** [comment]
@du00cs there's a few conflicts

File: docs/content/development/extensions-contrib/thrift.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+You may want to use another version of thrift, change the dependency in pom and compile yourself.
+
+## Thrift Parser
+
+
+| Field       | Type        | Description                              | Required |
+| ----------- | ----------- | ---------------------------------------- | -------- |
+| type        | String      | This should say `thrift`                 | yes      |
+| parseSpec   | JSON Object | Specifies the timestamp and dimensions of the data. Should be a Json parseSpec. | yes      |
+| thriftJar   | String      | path of thrift jar, if not provided, it will try to find the thrift class in classpath. Thrift jar in batch ingestion should be uploaded to HDFS first and configure `jobProperties` with `"tmpjars":"/path/to/your/thrift.jar"` | no       |
```
Comment (du00cs) [inline docs/content/development/extensions-contrib/thrift.md]: Adding thrift jar to middle manager's class path will indeed work for batch ingestion, but an HDFS or some path is much more flexible. You don't have to put a jar into each nodes. Maybe pull from somewhere by http is the best. In this way, batch and realtime ways are unified. I should try to make it.

**Comment by fjy** [comment]
👍

**Comment by justsmoke** [comment]
@du00cs I am using this extension with tranquility, but get the following exception: java.lang.IllegalArgumentException: requirement failed: Expected ByteBufferInputRowParser, got[io.druid.data.input.thrift.ThriftInputRowParser]. It seems that tranquility only supports ByteBufferInputRowParser ?

**Comment by du00cs** [comment]
@justsmoke You are right. I submit a pull request at Sep. [[Fix] Remove data type and InputRowParser type bindings](https://github.com/druid-io/tranquility/pull/193/files) , may be some help.

**Pull request merged**